fs/nfs_common/nfsacl.c

```diff
@@ -418,3 +425,138 @@ bool nfs_stream_decode_acl(struct xdr_stream *xdr, unsigned int *aclcnt,
 	return true;
 }
 EXPORT_SYMBOL_GPL(nfs_stream_decode_acl);
+
+#if CONFIG_TRUENAS
+static int
+convert_to_nfs40_ace(u32 *xdrbuf, struct nfs4_ace *ace)
+{
+	int error = 0;
+	u32 iflag, id;
+
+	ace->type = ntohl(*(xdrbuf++));
+	if (ace->type > NFS4_ACE_ACCESS_DENIED_ACE_TYPE)
+		return -EINVAL;
+
+	ace->flag = ntohl(*(xdrbuf++));
+	iflag = ntohl(*(xdrbuf++));
+	ace->access_mask = ntohl(*(xdrbuf++)) & NFS4_ACE_MASK_ALL;
+	id = ntohl(*(xdrbuf++));
+
+	if (iflag & ACEI4_SPECIAL_WHO) {
+		switch (id) {
+		case ACE4_SPECIAL_OWNER:
+			ace->whotype = NFS4_ACL_WHO_OWNER;
+			break;
+		case ACE4_SPECIAL_GROUP:
+			ace->whotype = NFS4_ACL_WHO_GROUP;
+			break;
+		case ACE4_SPECIAL_EVERYONE:
+			ace->whotype = NFS4_ACL_WHO_EVERYONE;
+			break;
+		}
+	} else {
+		ace->whotype = NFS4_ACL_WHO_NAMED;
+		if (ace->flag & NFS4_ACE_IDENTIFIER_GROUP) {
+			ace->who_gid = make_kgid(&init_user_ns, id);
+			if (!gid_valid(ace->who_gid)) {
+				error = -EINVAL;
+			}
+		} else {
+			ace->who_uid = make_kuid(&init_user_ns, id);
+			if (!uid_valid(ace->who_uid)) {
+				error = -EINVAL;
+			}
+		}
+	}
+
+	return error;
+}
+
+int
+convert_nfs41xdr_to_nfs40_acl(u32 *xdrbuf, size_t remaining, struct nfs4_acl *acl)
+{
+	int error = 0;
+	int i;
+
+	for (i = 0; i < acl->naces; i++, xdrbuf += NACE41_LEN) {
+		if (remaining < (NACE41_LEN * sizeof(u32))) {
```

Comment:
Just a small nit that you definitely don't need to fix but I'll make a comment nonetheless. You could calculate `(NACE41_LEN * sizeof(u32))` only once on the stack instead of 2 * acl->naces.

Reply:
Turns out we _already_ had a `#define` for this, so it certainly makes sense to use it here.  Thanks!

Reply:
Missed one more reference in that same function